### ISSUE TITLE
allow setting receive buffer size

### ISSUE THREAD

**Issue by mildsunrise** [versatica/mediasoup#554]
## Feature Request

I see no way to set the UDP receive buffer size (i.e. SO_RCVBUF), which can be a problem when trying to ingest 'bursty' streams into a router. Would it make sense to add it?

**Comment by jmillan** [comment]
Hi Alba,

Please use the forum for questions.

Regarding to your comment, we don't expose it but you can change the default send|recv buffer size value in the OS where you run mediasoup.

**Comment by ibc** [comment]
Yes @mildsunrise, please raise your question in the forum: https://mediasoup.discourse.group/.
There may be a chance for this feature request to be useful (other than setting UDP send|recv buffer size at OS level) and I'd like it to be discussed with more folks in the forum.

**Comment by mildsunrise** [comment]
This is not a question, it's a feature request and I was asking if a PR would be accepted.
From @ibc's response I understand it would?

Yes, I'm aware it can be changed at the OS level, but I don't think this is reasonable in general as it would affect all processes, and often isn't an option.

**Comment by ibc** [comment]
Hi @mildsunrise:

Yes, a PR may make sense. Just a few points about it:

- Would it work just for *nix? or also for Windows?
- libuv provides an API to set recv buffer size so it should be used: http://docs.libuv.org/en/v1.x/handle.html?highlight=uv_recv_buffer_size#c.uv_send_buffer_size
- How is the API gonna look like? is it gonna be a per Worker settings (so given recv size would be applied to all UDP/TCP sockets in all `Transports`)? or should we allow granular control and allow each port bind by a `Transport` to have its own recv size?
- Not mandatory, but just wondering if (once points above are clarified), you can also apply the new API in mediasoup Rust.

**Comment by penguinol** [comment]
We've achieved this in our own branch.
It's useful, case if the buffer is too small, the kernel may drop packets.
Every transport should have it's own buffer size, case the bandwidth is different for each transport.
We calcaute the size of buffer by the bandwidth of the transport.
PipeTransport usully needs a larger buffer size casue it will send/recv more steams than webrtctrasport.

**Comment by mildsunrise** [comment]
> Would it work just for *nix? or also for Windows?

Unless I'm missing something, libuv should support this option on Windows as well.

> How is the API gonna look like? is it gonna be a per Worker settings (so given recv size would be applied to all UDP/TCP sockets in all Transports)? or should we allow granular control and allow each port bind by a Transport to have its own recv size?

I think there's no reason to make it global (i.e. per-worker), my idea was a per-transport option.

> Not mandatory, but just wondering if (once points above are clarified), you can also apply the new API in mediasoup Rust.

Sure, looks easy enough. I'm not sure when I'll have time to work on the PR, but I'll keep it in mind.

also, cross linking: https://mediasoup.discourse.group/t/exposing-so-rcvbuf-so-sndbuf/2784

**Comment by vpalmisano** [comment]
It would be a nice feature to have. The only problem I see is that the Kernel imposes a maximum limit to the buffer size (ref. https://man7.org/linux/man-pages/man7/socket.7.html). For Linux these limits are specified in `/proc/sys/net/core/rmem_max` and `/proc/sys/net/core/wmem_max`. In the case the process is running with CAP_NET_ADMIN privilege, we could use the SO_RCVBUFFORCE and SO_SNDBUFFORCE variants.
I see that libuv is using this function (https://github.com/libuv/libuv/blob/master/src/uv-common.c#L587) that is a wrapper of `setsockopt`. I don't know if it is possible to use SO_RCVBUFFORCE/SO_SNDBUFFORCE without libuv modifications.

**Comment by ibc** [comment]
Definitely I don't want to modify libuv. If the kernel imposes limits, the user should increase those limits at kernel level.

**Comment by mildsunrise** [comment]
I agree. what [applications](https://github.com/FFmpeg/FFmpeg/blob/c122a6c10dedc05933a8b11a3b82f18afc8ca5b7/libavformat/udp.c#L865) [typically](https://github.com/GStreamer/gst-plugins-good/blob/master/gst/udp/gstmultiudpsink.c) [do](https://github.com/ossrs/srs/blob/47a7a5f82574caa2bbe2847698e8e2a7dccdc00a/trunk/src/app/srs_app_recv_thread.cpp#L551) is `setsockopt` the desired buffer size, then check with `getsockopt` that it was actually set. otherwise warn the user

**Comment by ibc** [comment]
The Transport constructor (in C++ side) can throw in that case so we are done.

**Comment by ibc** [comment]
Closing due to no activity.

**Comment by jbaudanza** [comment]
I would also like the ability to set buffer sizes on individual transports. 

Our server users 8-16 WebRTCServer objects. Since these handle traffic for all WebRTCTraffic, they need large buffers. We use the `net.core.rmem_default` kernel option to set this.

We also have hundreds of PlainRTPTransports that send the data to ffmpeg processes running internal on our network. These don't need to have the same large buffers as the WebRTCServer objects.

Let me know if I can prepare a PR for this.

**Comment by ibc** [comment]
> Let me know if I can prepare a PR for this.

Yes, but first I'd appreciate if you write here how the API is gonna look. Also, it must be implemented in both layers Node and Rust and must consider all supported OS and not just Linux.

**Comment by jbaudanza** [comment]
> Yes, but first I'd appreciate if you write here how the API is gonna look. Also, it must be implemented in both layers Node and Rust and must consider all supported OS and not just Linux.

I'd like to add the following options to `PlainTransportOptions`, `WebRtcTransportOptions`, and `WebRtcServerListenInfo`
```
sendBufferSize?: number;
recvBufferSize?: number;
```

These will be passed to `uv_send_buffer_size`, `uv_recv_buffer_size` on the underlaying UDP socket.

I'm not sure about the TCP sockets though. I personally don't have any need to adjust the buffer sizes on UDP sockets, but maybe this should be done for consistency. What do you think?

**Comment by ibc** [comment]
According to libuv docs that function works for both UDP and TCP. Anyway, let's please implement this once we finish the flatbuffer task (another ongoing and large PR) since this feature will require an additional channel request and we prefer to add it when the new channel mechanism is merged in v3 branch.

Reopening ticket.

**Comment by ibc** [comment]
Draft PR here: https://github.com/versatica/mediasoup/pull/1052, please provide feedback.

**Comment by ibc** [comment]
This is being implemented in PR #1084

**Comment by ibc** [comment]
PR https://github.com/versatica/mediasoup/pull/1084 has been merged in `flatbuffers` branch so this issue will be auto-closed once that branch is merged into `v3` branch.
